src/lets_be_rational/black.rs

```diff
@@ -0,0 +1,130 @@
+
+#[rustfmt::skip]
+fn small_t_expansion_of_normalised_black_call(h: f64, t: f64) -> f64 {
+    let a = 1.0 + h * (0.5 * SQRT_TWO_PI) * erfcx(-FRAC_1_SQRT_2 * h);
+    let w = t * t;
+    let h2 = h * h;
+
+    let expansion = 2.0 * t * (a + w * ((-1.0 + 3.0 * a + a * h2) / 6.0
+        + w * ((-7.0 + 15.0 * a + h2 * (-1.0 + 10.0 * a + a * h2)) / 120.0
+        + w * ((-57.0 + 105.0 * a + h2 * (-18.0 + 105.0 * a + h2 * (-1.0 + 21.0 * a + a * h2))) / 5040.0
+        + w * ((-561.0 + 945.0 * a + h2 * (-285.0 + 1260.0 * a + h2 * (-33.0 + 378.0 * a + h2 * (-1.0 + 36.0 * a + a * h2)))) / 362880.0
+        + w * ((-6555.0 + 10395.0 * a + h2 * (-4680.0 + 17325.0 * a + h2 * (-840.0 + 6930.0 * a + h2 * (-52.0 + 990.0 * a + h2 * (-1.0 + 55.0 * a + a * h2))))) / 39916800.0
+        + w * ((-89055.0 + 135135.0 * a + h2 * (-82845.0 + 270270.0 * a + h2 * (-20370.0 + 135135.0 * a + h2 * (-1926.0 + 25740.0 * a + h2 * (-75.0 + 2145.0 * a + h2 * (-1.0 + 78.0 * a + a * h2)))))) / 6227020800.0)))))));
+
+    let b = ONE_OVER_SQRT_TWO_PI * (-0.5 * (h * h + t * t)).exp() * expansion;
+    b.abs().max(0.0)
+}
+
+pub fn normalised_black_call_using_erfcx(h: f64, t: f64) -> f64 {
+    let b = 0.5
+        * (-0.5 * (h * h + t * t)).exp()
+        * (erfcx(-FRAC_1_SQRT_2 * (h + t)) - erfcx(-FRAC_1_SQRT_2 * (h - t)));
+    b.abs().max(0.0)
+}
+
+pub(crate) fn normalised_black_call(x: f64, s: f64) -> f64 {
+    if x > 0.0 {
+        return normalised_intrinsic(x, OptionType::Call) + normalised_black_call(-x, s);
+        // In the money.
+    }
+    if s <= x.abs() * DENORMALISATION_CUTOFF {
+        return normalised_intrinsic(x, OptionType::Call); // sigma=0 -> intrinsic value.
+    }
+
+    // Denote h := x/s and t := s/2.
+    // We evaluate the condition |h|>|η|, i.e., h<η  &&  t < τ+|h|-|η|  avoiding any divisions by s , where η = asymptotic_expansion_accuracy_threshold  and τ = small_t_expansion_of_normalised_black_threshold .
+    if x < s * ASYMPTOTIC_EXPANSION_ACCURACY_THRESHOLD
+        && 0.5 * s * s + x
+            < s * (SMALL_T_EXPANSION_OF_NORMALISED_BLACK_THRESHOLD
+                + ASYMPTOTIC_EXPANSION_ACCURACY_THRESHOLD)
+    {
+        return asymptotic_expansion_of_normalised_black_call(x / s, 0.5 * s);
+    }
+    if 0.5 * s < SMALL_T_EXPANSION_OF_NORMALISED_BLACK_THRESHOLD {
+        return small_t_expansion_of_normalised_black_call(x / s, 0.5 * s);
+    }
+
+    normalised_black_call_with_optimal_use_of_codys_functions(x, s)
+}
+
+pub(crate) fn normalised_black(x: f64, s: f64, q: f64) -> f64 {
+    normalised_black_call(if q < 0.0 { -x } else { x }, s) /* Reciprocal-strike call-put equivalence */
+}
+
+#[rustfmt::skip]
+pub fn asymptotic_expansion_of_normalised_black_call(h: f64, t: f64) -> f64 {
+    let e = (t / h) * (t / h);
+    let r = (h + t) * (h - t);
+    let q = (h / r) * (h / r);
+
+    // 17th order asymptotic expansion of A(h,t) in q
+    let asymptotic_expansion_sum = 2.0 + q * (-6.0 - 2.0 * e + 3.0 * q * (10.0 + e * (20.0 + 2.0 * e)
```

Comment:
I don't think so it is good idea  to use predefined values. @hayden4r4 what do you think about Newton-Raphson method to use here?

Reply:
I was using newton raphson originally but moved to this for reasons I can't remember, I am fine with moving it back to newton raphson to give an easier to understand/debug alternative to lets be rational

Reply:
while this makes for ugly code I'm a fan of what gives us better performance here, I'm curious what the added runtime cost of using newton raphson here would be